scipy/stats/stats.py

```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    # calculate MGC stat
+    stat, stat_dict = _mgc_stat(x, y, compute_distance)
+    stat_mgc_map = stat_dict["stat_mgc_map"]
+    opt_scale = stat_dict["opt_scale"]
+
+    # calculate permutation MGC p-value
+    pvalue, null_dist = _perm_test(x, y, stat, compute_distance, reps=reps,
+                                   workers=workers)
+
+    # save all stats (other than stat/p-value) in dictionary
+    mgc_dict = {"mgc_map": stat_mgc_map,
+                "opt_scale": opt_scale,
+                "null_dist": null_dist}
+
+    return MGCResult(stat, pvalue, mgc_dict)
+
+
+def _mgc_stat(x, y, compute_distance):
+    r"""
+    Helper function that calculates the MGC stat. See above for use.
+
+    Parameters
+    ----------
+    x, y : ndarray
+        `x` and `y` have shapes `(n, p)` and `(n, q)` or `(n, n)` and `(n, n)`
+        if distance matrices.
+    compute_distance : callable
+        A function that computes the distance or similarity among the samples
+        within each data matrix. Set to `None` if `x` and `y` are already
+        distance.
+
+    Returns
+    -------
+    stat : float
+        The sample MGC test statistic within `[-1, 1]`.
+    stat_dict : dict
+        Contains additional useful additional returns containing the following
+        keys:
+            - stat_mgc_map : ndarray
+                MGC-map of the statistics.
+            - opt_scale : (float, float)
+                The estimated optimal scale as a `(x, y)` pair.
+    """
+    # set distx and disty to x and y when compute_distance = None
+    distx = x
+    disty = y
+
+    if compute_distance is not None:
+        # compute distance matrices for x and y
+        distx = compute_distance(x)
+        disty = compute_distance(y)
+
+    # calculate MGC map and optimal scale
+    stat_mgc_map = _local_correlations(distx, disty, global_corr='mgc')[0]
+
+    n, m = stat_mgc_map.shape
+    if m == 1 or n == 1:
+        stat = stat_mgc_map[m - 1][n - 1]
+        opt_scale = m * n
+    else:
+        samp_size = len(distx) - 1
+
+        # threshold to find connected region of significant local correlations
+        sig_connect = _threshold_mgc_map(stat_mgc_map, samp_size)
+
+        # maximum within the significant region
+        stat, opt_scale = _smooth_mgc_map(sig_connect, stat_mgc_map)
+
+    stat_dict = {"stat_mgc_map": stat_mgc_map,
+                 "opt_scale": opt_scale}
+
+    return stat, stat_dict
+
+
+def _threshold_mgc_map(stat_mgc_map, samp_size):
+    r"""
+    Finds a connected region of significance in the MGC-map by thresholding.
+
+    Parameters
+    ----------
+    stat_mgc_map : ndarray
+        All local correlations within `[-1,1]`.
+    samp_size : int
+        The sample size of original data.
+
+    Returns
+    -------
+    sig_connect : ndarray
+        A binary matrix with 1's indicating the significant region.
+    """
+    m, n = stat_mgc_map.shape
+
+    # parametric threshold, based on a threshold is estimated based on the
```

Comment:
this sentence needs a rephrase.

also why the 0.02 value? give a reference if there's a justification somewhere (I assume there is)

Reply:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0